Plugins/org.mitk.lancet.jakarobot/src/internal/RobotArmRegistrationTab.cpp

```diff
@@ -0,0 +1,368 @@
+	 double Angle = -static_cast<double>(intValue);
+	 double templeArray[3] = { 0.0, 1, 0.0 };
+	 m_Robot->Rotate(templeArray, Angle);
+
+
+
+}
+void RobotArmRegistrationTab::rzm()
+{
+	int intValue = m_ui.lineEdit_intuitiveValue_degree_2->text().toInt();
+	double Angle = -static_cast<double>(intValue);
+	double templeArray[3] = { 0.0, 0.0, 1 };
+	m_Robot->Rotate(templeArray, Angle);
+
+	
+}
+
+void RobotArmRegistrationTab::captureRobot()
+{
+	m_ui.textBrowser_hans->append(QString("captureRobot"));
+
+	if (m_RobotRegistration.PoseCount() < 5) //The first five translations, 
+	{
+		CapturePose(true);
+	}
+	else if (m_RobotRegistration.PoseCount() >= 5 && m_RobotRegistration.PoseCount() < 10)
+	{
+		CapturePose(false);
+	}
+	else
+	{
+		vtkNew<vtkMatrix4x4> robotEndToFlangeMatrix;
+		m_RobotRegistration.GetTCPmatrix(robotEndToFlangeMatrix);
+
+		robotEndToFlangeMatrix->Invert();
+		m_ui.textBrowser_hans->append("Registration RMS: " + QString::number(m_RobotRegistration.RMS()));
+		std::cout << "Registration RMS: " << m_RobotRegistration.RMS() << std::endl;
+	}
+
+	m_ui.lineEdit_collectedRoboPose_2->setText(QString::number(m_RobotRegistration.PoseCount()));
+
+}
+
+void RobotArmRegistrationTab::CapturePose(bool translationOnly)
+{
+	m_ui.textBrowser_hans->append(QString("CapturePose"));
+	//get T_BaseToFlanger
+	vtkSmartPointer<vtkMatrix4x4>T_BaseToFlange= m_Robot->GetBaseToFlange();
+	//get T_CameraToEnd
+	vtkSmartPointer<vtkMatrix4x4>T_CameraToEnd= m_Camera->GetToolMatrixByName("RobotEndRF");
+	//get T_BaseRFToCamera
+	vtkSmartPointer<vtkMatrix4x4>T_BaseRFToCamera;
+	vtkMatrix4x4::Invert(m_Camera->GetToolMatrixByName("RobotBaseRF"),T_BaseRFToCamera);
+	//get T_BaseRFToEnd
+	vtkSmartPointer<vtkMatrix4x4> T_BaseRFToEnd;
+	
+	vtkMatrix4x4::Multiply4x4(T_BaseRFToCamera, T_CameraToEnd, T_BaseRFToEnd);
+	//Robotic arm registration
+	m_RobotRegistration.AddPoseWithVtkMatrix(T_BaseToFlange, T_BaseRFToEnd, translationOnly);
+}
+
+void RobotArmRegistrationTab::waitMove()
+{
+	m_ui.textBrowser_hans->append(QString("waitMove"));
+	bool bDone = false;
+	while (!bDone)
+	{
+		int nRet = HRIF_IsMotionDone(0, 0, bDone);
+		if (bDone == false)
+		{
+			QThread::msleep(10);
+			QApplication::processEvents();
+			std::cout << "is still moving" << std::endl;
+		}
+		else
+		{
+			int ret = HRIF_GrpStop(0, 0);
+			m_ui.textBrowser_hans->append(QString("robot is stop please continue"));
+			return;
+		}
+	}
+}
+
+void RobotArmRegistrationTab::autoCollection()
+{
+	m_ui.textBrowser_hans->append(QString("autoCollection"));
+	
+	if (HRIF_IsConnected(0))
+	{
+		m_ui.textBrowser_hans->append("Starting automatic registration!");
+		int moveCount = 1;
+		m_RobotRegistration.RemoveAllPose();
+		if (isAutoCollectionFlag)
+		{
+			m_Robot->SetTCPToFlange();
+			isAutoCollectionFlag = FALSE;
+		}
+		while (moveCount <= 10)
+		{
+			m_Robot->GoToInitialPos();
+			waitMove();
+			switch (moveCount) {
+			case 1:
+				xp();
```

Comment:
move waitMove();
				QThread::msleep(200);

				captureRobot(); out this swith